types/wlr_surface_layers.c

```diff
@@ -0,0 +1,212 @@
+#include <assert.h>
+#include <stdlib.h>
+#include <wayland-server.h>
+#include <wlr/types/wlr_output.h>
+#include <wlr/types/wlr_surface.h>
+#include <wlr/types/wlr_surface_layers.h>
+#include <wlr/util/log.h>
+#include "surface-layers-protocol.h"
+
+static const char *surface_layers_role = "layer_surface";
+
+static void layer_surface_set_interactivity(struct wl_client *client,
+		struct wl_resource *resource, uint32_t input_types,
+		uint32_t exclusive_types) {
+	struct wlr_layer_surface *surface = wl_resource_get_user_data(resource);
+	surface->input_types = input_types;
+	surface->exclusive_types = exclusive_types;
+	wl_signal_emit(&surface->events.set_interactivity, surface);
+}
+
+static void layer_surface_set_anchor(struct wl_client *client,
+		struct wl_resource *resource, uint32_t anchor) {
+	struct wlr_layer_surface *surface = wl_resource_get_user_data(resource);
+	surface->anchor = anchor;
+	wl_signal_emit(&surface->events.set_anchor, surface);
+}
+
+static void layer_surface_set_exclusive_zone(struct wl_client *client,
+		struct wl_resource *resource, uint32_t zone) {
+	struct wlr_layer_surface *surface = wl_resource_get_user_data(resource);
+	surface->exclusive_zone = zone;
+	wl_signal_emit(&surface->events.set_exclusive_zone, surface);
+}
+
+static void layer_surface_set_margin(struct wl_client *client,
+		struct wl_resource *resource, uint32_t horizontal,
+		uint32_t vertical) {
+	struct wlr_layer_surface *surface = wl_resource_get_user_data(resource);
+	surface->margin_horizontal = horizontal;
+	surface->margin_vertical = vertical;
+	wl_signal_emit(&surface->events.set_margin, surface);
+}
+
+void wlr_layer_surface_get_position(struct wlr_layer_surface *layer_surface,
+		double *x, double *y) {
```

Comment:
I think we should allow the layers to "stack" on top of each other (similar to X11 `STRUT_PARTIAL`). So if two layers both request to be at the bottom of the screen (like two bars), they will appear on top of each other without the clients having to know each others positions.

Reply:
So for instance, if I am desktop icons, I will want to put my icons in the top left corner but in a position relative to the bar height.

Reply:
No, that's intentionally not part of the design (for now). But that's somewhat possible with exclusivity - the idea is that a bar or something similar can set its exclusive zone to the top 10 pixels of the screen, then anything at or above that layer will not be shown on top of it (e.g. notifications).